src/newgrf_badge.cpp

```diff
@@ -0,0 +1,564 @@
+
+	ScopeResolver *GetScope(VarSpriteGroupScope scope = VSG_SCOPE_SELF, uint8_t relative = 0) override
+	{
+		switch (scope) {
+			case VSG_SCOPE_SELF: return &this->self_scope;
+			default: return ResolverObject::GetScope(scope, relative);
+		}
+	}
+
+	GrfSpecFeature GetFeature() const override;
+	uint32_t GetDebugID() const override;
+};
+
+GrfSpecFeature BadgeResolverObject::GetFeature() const
+{
+	return GSF_BADGES;
+}
+
+uint32_t BadgeResolverObject::GetDebugID() const
+{
+	return this->self_scope.badge.index.base();
+}
+
+/**
+ * Constructor of the badge resolver.
+ * @param badge Badge being resolved.
+ * @param feature GRF feature being used.
+ * @param introduction_date Optional introduction date of entity.
+ * @param callback Callback ID.
+ * @param callback_param1 First parameter (var 10) of the callback.
+ * @param callback_param2 Second parameter (var 18) of the callback.
+ */
+BadgeResolverObject::BadgeResolverObject(const Badge &badge, GrfSpecFeature feature, std::optional<TimerGameCalendar::Date> introduction_date, CallbackID callback, uint32_t callback_param1, uint32_t callback_param2)
+		: ResolverObject(badge.grf_prop.grffile, callback, callback_param1, callback_param2), self_scope(*this, badge, introduction_date)
+{
+	assert(feature <= GSF_END);
+	if (this->self_scope.badge.grf_prop.spritegroup[feature] == nullptr) feature = GSF_END;
+	this->root_spritegroup = badge.grf_prop.spritegroup[feature];
+}
+
+/**
+ * Test for a matching badge in a list of badges, returning the number of matching bits.
+ * @param grffile GRF file of the current varaction.
+ * @param badges List of badges to test.
+ * @param parameter GRF-local badge index.
+ * @returns true iff the badge is present.
+ */
+uint32_t GetBadgeVariableResult(const GRFFile &grffile, std::span<const BadgeID> badges, uint32_t parameter)
+{
+	if (parameter >= std::size(grffile.badge_list)) return UINT_MAX;
+
+	BadgeID index = grffile.badge_list[parameter];
+	return std::ranges::find(badges, index) != std::end(badges);
+}
+
+/**
+ * Mark a badge a seen (used) by a feature.
+ */
+void MarkBadgeSeen(BadgeID index, GrfSpecFeature feature)
+{
+	Badge *b = GetBadge(index);
+	assert(b != nullptr);
+	b->SetFeature(feature);
+}
+
+/**
+ * Append copyable badges from a list onto another.
+ * Badges must exist and be marked with the Copy flag.
+ * @param dst Destination badge list.
+ * @param src Source badge list.
+ * @param feature Feature of list.
+ */
+void AppendCopyableBadgeList(std::vector<BadgeID> &dst, std::span<const BadgeID> src, GrfSpecFeature feature)
+{
+	for (const BadgeID &index : src) {
+		/* Is badge already present? */
+		if (std::ranges::find(dst, index) != std::end(dst)) continue;
+
+		/* Is badge copyable? */
+		Badge *badge = GetBadge(index);
+		if (badge == nullptr) continue;
+		if ((badge->flags & BadgeFlags::Copy) == BadgeFlags::None) continue;
```

Comment:
Didn't you introduce `HasFlag` for this? Or rather `!HasFlag` in this situation.